executor/analyze.go

```diff
@@ -1370,13 +1374,19 @@ workLoop:
 			var err error
 			idx := e.indexes[task.slicePos-colLen]
 			sampleItems := make([]*statistics.SampleItem, 0, task.rootRowCollector.Base().Samples.Len())
+		indexSampleCollectLoop:
 			for _, row := range task.rootRowCollector.Base().Samples {
 				if len(idx.Columns) == 1 && row.Columns[idx.Columns[0].Offset].IsNull() {
 					continue
 				}
 
 				b := make([]byte, 0, 8)
 				for _, col := range idx.Columns {
+					// If the index value contains one value which is too long, we think that it's a value that doesn't occur many times.
+					if (col.Length != types.UnspecifiedLength && col.Length > statistics.MaxSampleValueLength) ||
```

Comment:
Why do we check `col.Length`? Even if `col.Length > statistics.MaxSampleValueLength`, the length of the sample value may still be small. Besides, we can move the check out of the sample loop.